### PR TITLE
Fix QNDF

### PR DESCRIPTION
**Pull request by JunpengGao233**
The QNDF in master branch fails to solve the Lorenz system. This PR fixes various errors in the QNDF implementation. After the fixes, both in-place and out-of-place QNDF implementations are able to solve the Lorenz system within 5000 timesteps.

Here is the plot:
lorenz

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Can you resolve the conflicts with the controller updates?

File: src/nlsolve/newton.jl [hunk 1]
```diff
@@ -97,6 +98,7 @@ Equations II, Springer Series in Computational Mathematics. ISBN
 
   atmp = calculate_residuals(dz, uprev, ustep, opts.abstol, opts.reltol, opts.internalnorm, t)
   ndz = opts.internalnorm(atmp, t)
+  #@show ndz
```
Comment (YingboMa) [inline src/nlsolve/newton.jl:101]: Remove this

File: src/nlsolve/newton.jl [hunk 2]
```diff
@@ -52,6 +52,7 @@ Equations II, Springer Series in Computational Mathematics. ISBN
   @unpack z,tmp,γ,α,cache = nlsolver
   @unpack tstep,W,invγdt = cache
 
+
```
Comment (YingboMa) [inline src/nlsolve/newton.jl:55]: Same here

File: src/nlsolve/nlsolve.jl [hunk 3]
```diff
@@ -54,6 +54,7 @@ function nlsolve!(nlsolver::AbstractNLSolver, integrator, cache=nothing, repeat_
     iter > 1 && (η = θ / (1 - θ))
     if (iter == 1 && ndz < 1e-5) || (iter > 1 && (η >= zero(η) && η * ndz < κ))
       nlsolver.status = Convergence
+      #@info "converge"
```
Comment (YingboMa) [inline src/nlsolve/nlsolve.jl:57]: here

**Review by YingboMa — APPROVED** [review]
Looks right to me.

**Comment by YingboMa** [comment]
This is basically a complete rewrite of QNDF. Good work!

File: src/bdf_utils.jl [hunk 4]
```diff
@@ -1,72 +1,23 @@
...
+  @views @.. D[:,k+1] = dd
+  for i in k:-1:1
+    @views @.. D[:,i] = D[:,i] + D[:,i+1]
   end
+  return nothing
 end
 
-global const γₖ = @SVector[sum(1//j for j in 1:k) for k in 1:6]
\ No newline at end of file
+global const γₖ = @SVector[sum(1//j for j in 1:k) for k in 1:6]
```
Comment (devmotion) [inline src/bdf_utils.jl:23]: I don't think `global const` is a thing?
```suggestion
const γₖ = @SVector[sum(1//j for j in 1:k) for k in 1:6]
```
One could also hardcode the values I guess:
```suggestion
# sum(1//j for j in 1:k) for k in 1:6
const γₖ = @SVector[1//1, 3//2, 11//6, 25//12, 137//60, 49//20]
```

File: src/bdf_utils.jl [hunk 5]
```diff
@@ -69,4 +70,25 @@ function reinterpolate_history!(cache::OrdinaryDiffEqConstantCache, D, R, k)
...
+  @inbounds for r = 1:k
+    R[1,r] = -r * ρ
+    for j = 2:k
+      R[j,r] = R[j-1,r] * ((j-1) - r * ρ)/j
+    end
+  end
+  SArray(R)
+end
+
+function update_D!(D, dd, k)
```
Comment (YingboMa) [inline src/bdf_utils.jl:85]:
```suggestion
function update_D!(D, dd, k)
  dd = _vec(dd)
```

**Comment by YingboMa** [comment]
I ran the POLLU benchmark with the new QNDF

pollu_bench

File: src/bdf_utils.jl [hunk 6]
```diff
@@ -69,4 +70,26 @@ function reinterpolate_history!(cache::OrdinaryDiffEqConstantCache, D, R, k)
...
+  dd = _vec(dd)
+  @views @.. D[:,k+2] = dd - D[:,k+1]
+  @views @.. D[:,k+1] = dd
+  for i in k:-1:1
+    @views @.. D[:,i] = D[:,i] + D[:,i+1]
+  end
+  return nothing
+end
+
+const γₖ = @SVector[sum(1//j for j in 1:k) for k in 1:6]
```
Comment (devmotion) [inline src/bdf_utils.jl:95]: Actually, due to https://github.com/JunpengGao233/OrdinaryDiffEq.jl/blob/cc45e743e9505809d268c9c2d4ecd5344eba80be/src/caches/bdf_caches.jl#L290 and a similar line for the mutable cache, I am wondering if this line is needed at all.

Comment (YingboMa) [reply src/bdf_utils.jl:95]: This is needed for the old QNDF1 and QNDF2 methods

**Comment by YingboMa** [comment]
The test failures seem unrelated.

**Comment by YingboMa** [comment]
QNDF is actually the best solver for BCR now.

bcr_bench

**Comment by JunpengGao233** [comment]
Thank you for helping make benchmarks. The bdf_utils.jl has some functions and constants that QNDF1 and QNDF2 are dependent so I recover them later. If needed, I could modify these methods later. 

Thanks for the instructions and help of Yingbo.

**Comment by ChrisRackauckas** [comment]
ABDF2 is the other. Yeah, it would be good to get those cleaned up because they are common in PDE applications still.

**Comment by YingboMa** [comment]
I merged this PR since the test failures are unrelated. I will look into these.

**Comment by ChrisRackauckas** [comment]
bcr_bench_2
pollu_bench_2
bcr_step_control

**Comment by YingboMa** [comment]
The above plots are with an optimized QNDF. I have a few more improves locally. I will open a PR soon.

**Pull request merged**